Spigot-Server-Patches/0619-Added-SoundBuilder.patch

```diff
@@ -0,0 +1,438 @@
++     */
++    @Override
++    public PaperSoundBuilder play() {
++
++        if (this.location == null) {
++            throw new IllegalStateException("Please specify location for this sound");
++        }
++
++        if (this.category == null) {
++            throw new IllegalStateException("Please specify category for this sound");
++        }
++
++        if (receivers == null || receivers.isEmpty()) {
++            location.getWorld().playSound(location, sound, volume, pitch);
++        } else {
++            PacketPlayOutCustomSoundEffect customSoundEffect = new PacketPlayOutCustomSoundEffect(MinecraftKey.a(sound.getKey().getKey()), net.minecraft.server.SoundCategory.valueOf(category.name()), Vec3D.a(new BaseBlockPosition(location.getX(), location.getY(), location.getZ())), volume, pitch);
++            for (Player player : receivers) {
++                ((CraftPlayer) player).getHandle().playerConnection.sendPacket(customSoundEffect);
++            }
++        }
++
++        return this;
++    }
++
++    /**
++     * @return The sound going to be sent
++     */
++    @Override
++    public Sound sound() {
++        return sound;
++    }
++
++    /**
++     * Changes what sound will be sent
++     *
++     * @param sound The sound
++     * @return a reference to this object.
++     */
++    @Override
++    public PaperSoundBuilder sound(Sound sound) {
++        this.sound = sound;
++        return this;
++    }
++
++    /**
++     * @return The category under which the sound will be played
++     */
++    @Override
++    public SoundCategory category() {
++        return category;
++    }
++
++    /**
++     * Changes under which category the sound will be played
++     */
++    @Override
++    public PaperSoundBuilder category(SoundCategory category) {
++        this.category = category;
++        return this;
++    }
++
++    /**
++     * @return The volume level of the sound played
++     */
++    @Override
++    public Float volume() {
++        return volume;
++    }
++
++    /**
++     * Changes the volume level of the sound played
++     *
++     * @param volume The volume
++     * @return a reference to this object.
++     */
++    @Override
++    public PaperSoundBuilder volume(Float volume) {
++        this.volume = volume;
++        return this;
++    }
++
++    /**
++     * @return The pitch of the sound played
++     */
++    @Override
++    public Float pitch() {
++        return pitch;
++    }
++
++    /**
++     * Changes the pitch of the sound played
++     *
++     * @param pitch The pitch
++     * @return a reference to this object.
++     */
++    @Override
++    public PaperSoundBuilder pitch(Float pitch) {
++        this.pitch = pitch;
++        return this;
++    }
++
++    /**
++     * @return List of players who will receive the sound, or null for all in world
++     */
++    @Override
++    public List<Player> receivers() {
++        return Lists.newLinkedList(receivers);
```

Comment:
Why are you using linked lists? Makes no sense. An `ArrayList` is a better choice here.

Reply:
I personally would go for a Set especially a hash set on creation. Having receivers twice in the collection makes no sense.

Reply:
There is another comment thread about the return type. He wants to use a List because that's what Paper does everywhere according to him.